Add generators

crertel opened the pull request:
This PR is about adding `mix scenic.gen.scene` and `mix scenic.gen.component` helpers.

It's currently a work in progress, but if this is looking like a good idea we can wrap it up.

Current questions are:

* How to we get the proper base module name, e.g.  `MyApp` in `MyApp.Scenes.SceneName` when we've run `mix scenic.gen.scene scene_name`?
* Are there additional considerations around umbrella apps or whatever?

crertel committed:
Fix Elixir 1.11 gripes about missing eex.

crertel committed:
Add mix scenic.gen.scene first cut.

Eiji7 commented:
> How to we get the proper base module name, e.g. MyApp in MyApp.Scenes.SceneName when we've run mix scenic.gen.scene scene_name?

Simply look how `Phoenix` core team solves it … I recommend to read especially `Mix.Phoenix` code.

Here you have app `module` name:
https://github.com/phoenixframework/phoenix/blob/e05af0ad5527c9a192de122f3e43d7ed4c7a4c9f/lib/mix/phoenix.ex#L123-L150

Here you have `web` path:
https://github.com/phoenixframework/phoenix/blob/e05af0ad5527c9a192de122f3e43d7ed4c7a4c9f/lib/mix/phoenix.ex#L196-L204

> Are there additional considerations around umbrella apps or whatever?

https://github.com/phoenixframework/phoenix/blob/e05af0ad5527c9a192de122f3e43d7ed4c7a4c9f/lib/mix/tasks/phx.gen.schema.ex#L102-L104

Short: don't complicate and simply raise :smiling_imp: 

If you really want you can add `app` name option.

crertel committed:
Fix root module lookup, add umbrella check.

crertel commented:
@Eiji7 went ahead and added that, and made some common helpers based on the Phoenix approach.

It'd still be cool to get some feedback from @boydm on this spike. :)

crertel committed:
Cleanup naming, holdovers from component/scene split.

crertel committed:
Add scenic.gen.component mix task.

crertel committed:
Fix test setup for scene, fix module naming for components.

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,74 @@
+
+  # --------------------------------------------------------
+  defp generate(component_filename, component_module) do
+    if Mix.Project.umbrella?() do
+      Mix.raise "mix scenic.gen.component must be invoked from within your scenic application root directory."
+    end
```
I believe that this should be on the top of `run/1` function. Firstly everything made before does not makes sense if we want to raise. Secondly it would be inconsistent and therefore confusing to display for example `help` message and when passing correct arguments `raise`.

crertel replied:
Fixed!

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -0,0 +1,37 @@
+    @graph Graph.build()
+    |> text("", text_align: :center, translate: {100, 200}, id: :text)
+
+    # TODO: Explanation of info/1 callback here.
```
We do not need an extra documentation or comments if this is callback. In that case all we need is proper `@impl` and all the documentation in `behaviour` and optionally guides.

crertel replied:
Theoretically, yes. Sometimes it's helpful just to give people the reference documentation right where they would be running into it. Phoenix does this a lot in the gen'd code. I'm not super picky either way.

Eiji7 replied:
Yes, `phoenix` gives many comments, `@doc` and `@moduledoc` for generated modules. The `@impl My.Behaviour` is the definitely minimal we need, so it's enough (right now) just to make easy to understand for maintainers and contributors (otherwise people would spend time on looking for it in code source). Later we can freely focus on extra comments for new developers.

crertel replied:
Gotcha, went ahead and did that.

crertel committed:
Move umbrella checks to run/1.

crertel commented:
Thanks for looking over this with me @Eiji7 !

One bothersome detail is that the tests are failing for the `gen.scenic.scene`, because I don't think we're autodetecting the root module quite correctly. I've hit my spare cycles for this for today, but am open to ideas.

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,74 @@
+        Common.elixir_version_check!()
+        app = opts[:app] || Path.basename(Path.expand(component_name))
+        Common.check_application_name!(app, !opts[:app])
+        mod = opts[:module] || Macro.camelize(app)
```
Please correct me if I'm wrong, but typically community is doing the opposite of that (`underscore` module) and that's because:
```elixir
iex> Macro.underscore("MyHTTP")
"my_http"
iex> Macro.camelize("my_http")
"MyHttp"
```
see:
https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Channel.html#content
and other `mix` tasks in for this library.

crertel replied:
Oh, you mean, doing:

`mix gen.scenic.scene MyScene`

instead of 

`mix gen.scenic.scene my_scene`

?

Eiji7 replied:
exactly

crertel replied:
Good idea, will change.

crertel replied:
Fixed!

Eiji7 commented:
> @crertel One bothersome detail is that the tests are failing for the gen.scenic.scene, because I don't think we're autodetecting the root module quite correctly. I've hit my spare cycles for this for today, but am open to ideas.

You have 2 options:
1. Don't change generated app name (`phoenix` also checks for `PhoenixWeb` module name)
2. Change configuration like:

```elixir
Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))
```

Eiji7 commented on test/scenic_gen_scene_test.exs:
```diff
@@ -0,0 +1,63 @@
+
+  test "gen.scene" do
+    in_tmp("new with defaults", fn ->
+       assert capture_io(fn ->
+        Mix.Tasks.Scenic.New.run([@app_name])
+        Mix.Tasks.Scenic.Gen.Scene.run([@scene_name])
+
+        assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
+          assert file =~ "defmodule #{@module_name}.Component.#{@scene_module_name} do"
+        end)
+      end) =~ "Created component #{@scene_module_name}."
```
```suggestion
      func = fn ->
        Mix.Tasks.Scenic.New.run([@app_name])
        Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))
        Mix.Tasks.Scenic.Gen.Scene.run([@scene_name])

        assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
          assert file =~ "defmodule #{@module_name}.Component.#{@scene_module_name} do"
        end)
      end

      assert capture_io(func) =~ "Created component #{@scene_module_name}."
```

crertel replied:
I like it!

crertel committed:
Update test/scenic_gen_scene_test.exs

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Change scenic.gen.scene to take CamelCase mod name.

crertel committed:
Change scenic.gen.component to take ModuleName.

crertel committed:
Add scenic.gen.component test.

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,74 @@
+  @doc false
+  def run(argv) do
+    if Mix.Project.umbrella?() do
+      Mix.raise "mix scenic.gen.component must be invoked from within your scenic application root directory."
```
`Elixir`'s formatter wants to break it as this is too long. Since we are using `scenic.gen.component` string part at least 2 times in this file I suggest to make it an attribute and use it as:
```elixir
@task "scenic.gen.component"
"#{@task}"
```

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,74 @@
+    module: :string
+  ]
+
+
+  # --------------------------------------------------------
+  @doc false
+  def run(argv) do
+    if Mix.Project.umbrella?() do
+      Mix.raise "mix scenic.gen.component must be invoked from within your scenic application root directory."
+    end
+
+    {opts, argv} = OptionParser.parse!(argv, strict: @switches)
+
+    case argv do
+      [] ->
+        Mix.Tasks.Help.run(["scenic.gen.component"])
```
To satisfy `Elixir`'s formatter:
```suggestion
  @task "scenic.gen.component"

  # --------------------------------------------------------
  @doc false
  def run(argv) do
    if Mix.Project.umbrella?() do
      message = "mix #{@task} must be invoked from within your scenic application root directory."
      Mix.raise(message)
    end

    {opts, argv} = OptionParser.parse!(argv, strict: @switches)

    case argv do
      [] ->
        Mix.Tasks.Help.run([@task])
```

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,74 @@
+  # --------------------------------------------------------
+  defp generate(component_filename, component_module) do
+    assigns = [
+      app_module_name: Common.base_module(), #TOD: Figure out the right module name.
```
It has been already fixed, right?
```suggestion
      app_module_name: Common.base_module(),
```

crertel replied:
Fixed!

Eiji7 commented:
If you apply above suggestions to `lib/mix/tasks/scenic.gen.component.ex` and `lib/mix/tasks/scenic.gen.scene.ex ` and then call `mix format` everything should be ok.

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,74 @@
+    create_directory("lib/components")
+    create_file(component_path, component_new_template(assigns))
+
+    """
+    Created component #{component_filename}.
+    """
+    |> Mix.shell().info()
```
```suggestion
    Mix.shell().info("Created component #{component_filename}.")
```

For me such output looks good:
```elixir
$ mix scenic.gen.scene my_scene
* creating lib/scenes
* creating lib/scenes/my_scene.ex
Created scene lib/scenes/my_scene.ex.
```
previously:
```elixir
$ mix scenic.gen.scene my_scene
* creating lib/scenes
* creating lib/scenes/my_scene.ex
Created scene lib/scenes/my_scene.ex.
                                                               # <- newline here is not needed
```

crertel replied:
++

Eiji7 commented:
Before merge we would need a simple cleanup:
1. Minimal callback documentation like `@impl Scenic.Component` (as said rest could be added later, but this is must have)
2. Apply changes to pass `mix format` checks on Travis (already described above)
3. Remove `IO.inspect/1` call (added for debugging)
4. Remove commented tests code or remove comments and update tests
5. Following `phoenix` task file name should be `task.name.ex` for example: `lib/mix/tasks/scenic.gen.component.ex`
see: https://github.com/phoenixframework/phoenix/tree/master/lib/mix/tasks
6. Keep same file/directory structure in `lib` and `test`

`lib/mix/tasks/scenic_gen_component.ex` -> `lib/mix/tasks/scenic.gen.component.exs`
`test/scenic_gen_component_test.exs` -> `test/mix/tasks/scenic.gen.component_test.exs`

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Remove inspect spam.

crertel committed:
Mix format pass.

crertel committed:
Remove TODs.

crertel committed:
Merge branch 'add-generators' of github.com:crertel/scenic_new into add-generators

crertel commented:
Cool, did the cleanup except for a few things:

> Remove commented tests code or remove comments and update tests

Right, am waiting on feedback from @boydm if we're actually testing what we need to.

> Keep same file/directory structure in lib and test

I'm matching existing practices (more or less) in that directory. If we want to reorganize things, that should be a different PR.

> Following phoenix task file name should be task.name.ex for example: lib/mix/tasks/scenic.gen.component.ex

I did that with the two generators I've added; changing the existing stuff should be a PR in itself (see previous).

Mostly just waiting on feedback from Boyd now.

crertel committed:
Replace doc stubs with impls.

Eiji7 commented on test/scenic_gen_component_test.exs:
```diff
@@ -0,0 +1,66 @@
+  @app_name "scenic_demo"
+  @module_name "ScenicDemo"
+
+  test "gen.scene" do
+    Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))
+
+    in_tmp("new with defaults", fn ->
+      func = fn ->
+        Mix.Tasks.Scenic.New.run([@app_name])
+        Mix.Tasks.Scenic.Gen.Component.run([@component_module_name])
+
+        assert_file("lib/components/#{@component_name}.ex", fn file ->
+          assert file =~ "defmodule #{@module_name}.Components.#{@component_module_name} do"
+        end)
+      end
+
+      assert capture_io(func) =~ "Created component #{@component_module_name}."
+    end)
+  end
```
Just noticed that we can avoid one extra indention by simply moving `func` outside of `in_tmp/2` call:
```suggestion
  test "gen.scene" do
    Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))

    func = fn ->
      Mix.Tasks.Scenic.New.run([@app_name])
      Mix.Tasks.Scenic.Gen.Component.run([@component_module_name])

      assert_file("lib/components/#{@component_name}.ex", fn file ->
        assert file =~ "defmodule #{@module_name}.Components.#{@component_module_name} do"
      end)
    end

    in_tmp("new with defaults", fn ->
      assert capture_io(func) =~ "Created component #{@component_module_name}."
    end)
  end
```

Eiji7 commented:
@crertel Makes sense. I would create a PR with those and similar changes for whole project and also I would add credo dependency.

crertel committed:
Update test/scenic_gen_component_test.exs

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -6,18 +6,18 @@ defmodule <%= @app_module_name %>.Components.<%= @component_module_name %> do
     def verify( _ ), do: :invalid_data
 
-    # TODO: Explanation of init/2 callback here.
+    @impl Scenic.Component
```
I don't see that callback in documentation:
https://hexdocs.pm/scenic/Scenic.Component.html#summary

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -32,6 +32,6 @@ defmodule <%= @app_module_name %>.Components.<%= @component_module_name %> do
     end
 
-    # TODO Explanation of event handlers here.
+    @impl Scenic.Component
```
Same here

Eiji7 commented on test/scenic_gen_scene_test.exs:
```diff
@@ -0,0 +1,66 @@
+  @app_name "scenic_demo"
+  @module_name "ScenicDemo"
+
+  test "gen.scene" do
+    Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))
+
+    in_tmp("new with defaults", fn ->
+      func = fn ->
+        Mix.Tasks.Scenic.New.run([@app_name])
+        Mix.Tasks.Scenic.Gen.Scene.run([@scene_module_name])
+
+        assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
+          assert file =~ "defmodule #{@module_name}.Scene.#{@scene_module_name} do"
+        end)
+      end
+
+      assert capture_io(func) =~ "Created scene #{@scene_module_name}."
+    end)
+  end
```
Same as in `test/scenic_gen_component_test.exs`
```suggestion
  test "gen.scene" do
    Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))

    func = fn ->
      Mix.Tasks.Scenic.New.run([@app_name])
      Mix.Tasks.Scenic.Gen.Scene.run([@scene_module_name])

      assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
        assert file =~ "defmodule #{@module_name}.Scene.#{@scene_module_name} do"
      end)
    end

    in_tmp("new with defaults", fn ->
      assert capture_io(func) =~ "Created scene #{@scene_module_name}."
    end)
  end
```

crertel committed:
Remove incorrect @impls.

crertel committed:
Merge branch 'add-generators' of github.com:crertel/scenic_new into add-generators

crertel committed:
Update test/scenic_gen_scene_test.exs

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel commented:
For some of these, I think that (as we saw when I mistakenly added some imps to the component boilerplate) maybe we need more callbacks specified back in core scenic.

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -17,7 +17,6 @@ defmodule <%= @app_module_name %>.Components.<%= @component_module_name %> do
     def verify( text ) when is_bitstring(text), do: {:ok, text}
     def verify( _ ), do: :invalid_data
 
-    @impl Scenic.Component
```
Oh, I though that those should be added, but for `Scenic.Scene`. If remember correctly `use Scenic.Component` adds `use Scenic.Scene`.

crertel replied:
![image](https://user-images.githubusercontent.com/1707779/107070389-e50e0500-67a8-11eb-9640-6bc7a0e06001.png)

crertel replied:
(agreed)

Eiji7 commented on test/scenic_gen_component_test.exs:
```diff
@@ -0,0 +1,66 @@
+  import ScenicNew.MixHelper
+  import ExUnit.CaptureIO
+
+  @component_name "test_component"
+  @component_module_name "TestComponent"
+  @app_name "scenic_demo"
+  @module_name "ScenicDemo"
+
+  test "gen.scene" do
+    Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))
+
+    func = fn ->
+      Mix.Tasks.Scenic.New.run([@app_name])
+      Mix.Tasks.Scenic.Gen.Component.run([@component_module_name])
+
+      assert_file("lib/components/#{@component_name}.ex", fn file ->
+        assert file =~ "defmodule #{@module_name}.Components.#{@component_module_name} do"
+      end)
+    end
+
+    in_tmp("new with defaults", fn ->
+      assert capture_io(func) =~ "Created component #{@component_module_name}."
+    end)
+  end
```
```suggestion
  @component_name "test_component"
  @component_module_name "TestComponent"
  @app_name "scenic_new"
  @module_name "ScenicNew"

  test "gen.scene" do
    func = fn ->
      Mix.Tasks.Scenic.New.run([@app_name])
      Mix.Tasks.Scenic.Gen.Component.run([@component_module_name])

      assert_file("lib/components/#{@component_name}.ex", fn file ->
        assert file =~ "defmodule #{@module_name}.Component.#{@component_module_name} do"
      end)
    end

    in_tmp("new with defaults", fn ->
      assert capture_io(func) =~ "Created component #{@component_module_name}."
    end)
  end
```

Eiji7 commented on test/scenic_gen_scene_test.exs:
```diff
@@ -0,0 +1,66 @@
+  import ScenicNew.MixHelper
+  import ExUnit.CaptureIO
+
+  @scene_name "test_scene"
+  @scene_module_name "TestScene"
+  @app_name "scenic_demo"
+  @module_name "ScenicDemo"
+
+  test "gen.scene" do
+    Application.put_env(String.to_atom(@app_name), :namespace, String.to_atom(@module_name))
+
+    func = fn ->
+      Mix.Tasks.Scenic.New.run([@app_name])
+      Mix.Tasks.Scenic.Gen.Scene.run([@scene_module_name])
+
+      assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
+        assert file =~ "defmodule #{@module_name}.Scene.#{@scene_module_name} do"
+      end)
+    end
+
+    in_tmp("new with defaults", fn ->
+      assert capture_io(func) =~ "Created scene #{@scene_module_name}."
+    end)
+  end
```
```suggestion
  @scene_name "test_scene"
  @scene_module_name "TestScene"
  @app_name "scenic_new"
  @module_name "ScenicNew"

  test "gen.scene" do
    func = fn ->
      Mix.Tasks.Scenic.New.run([@app_name])
      Mix.Tasks.Scenic.Gen.Scene.run([@scene_module_name])

      assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
        assert file =~ "defmodule #{@module_name}.Scene.#{@scene_module_name} do"
      end)
    end

    in_tmp("new with defaults", fn ->
      assert capture_io(func) =~ "Created scene #{@scene_module_name}."
    end)
  end
```

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,71 @@
+  import Mix.Generator
+  alias ScenicNew.Common
+
+  @switches [
+    app: :string,
+    module: :string
+  ]
```
```suggestion
  @switches [app: :string, module: :string]
```

Eiji7 commented on lib/mix/tasks/scenic.gen.scene.ex:
```diff
@@ -0,0 +1,73 @@
+  import Mix.Generator
+  alias ScenicNew.Common
+
+  @switches [
+    app: :string,
+    module: :string
+  ]
```
```suggestion
  @switches [app: :string, module: :string]
```

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -0,0 +1,35 @@
+    def verify( text ) when is_bitstring(text), do: {:ok, text}
+    def verify( _ ), do: :invalid_data
+
+    def init( text, opts ) do
```
```suggestion
    @impl Scenic.Scene
    def init( text, opts ) do
```

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -0,0 +1,35 @@
+        {:ok, state, push: graph}
+    end
+
+    def handle_input(_event, _context, state), do: {:noreply, state}
```
```suggestion
    @impl Scenic.Scene
    def handle_input(_event, _context, state), do: {:noreply, state}
```

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  # --------------------------------------------------------
+  defp generate(component_filename, component_module) do
+    assigns = [
+      app_module_name: Common.base_module(),
+      component_module_name: component_module
+    ]
+
+    component_path = "lib/components/#{component_filename}.ex"
+
+    create_directory("lib/components")
+    create_file(component_path, component_new_template(assigns))
+
+    Mix.shell().info("Created component #{component_filename}.")
+  end
```
```suggestion
  defp generate(component_filename, component_module) do
```suggestion
    component_path = "lib/components/#{component_filename}.ex"
    create_directory("lib/components")
    create_file(component_path, component_new_template(component_module: component_module))
    Mix.shell().info("Created component #{component_module}.")
```
  end
```

Eiji7 commented on lib/mix/tasks/scenic.gen.scene.ex:
```diff
@@ -0,0 +1,73 @@
+  end
+
+  # --------------------------------------------------------
+  defp generate(scene_filename, scene_module) do
+    assigns = [
+      app_module_name: Common.base_module(),
+      scene_module_name: scene_module
+    ]
+
+    scene_path = "lib/scenes/#{scene_filename}.ex"
+
+    create_directory("lib/scenes")
+    create_file(scene_path, scene_new_template(assigns))
+
+    """
+    Created scene #{scene_path}.
+    """
+    |> Mix.shell().info()
+  end
```
```suggestion
  defp generate(scene_filename, scene_module) do
    scene_path = "lib/scenes/#{scene_filename}.ex"
    create_directory("lib/scenes")
    create_file(scene_path, scene_new_template(scene_module: scene_module))
    Mix.shell().info("Created scene #{scene_module}.")
  end
```

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,71 @@
+      [] ->
+        Mix.Tasks.Help.run([@task])
+
+      [component_module_name | _] ->
+        Common.elixir_version_check!()
+        file_name = opts[:module] || Macro.underscore(component_module_name)
+        mod_name = opts[:app] || component_module_name
+
+        Common.check_mod_name_validity!(mod_name)
+        Common.check_mod_name_availability!(mod_name)
+
+        generate(file_name, mod_name)
```
```suggestion
      [component_module | _] ->
        Common.elixir_version_check!()
        file_name = Macro.underscore(opts[:module] || component_module)
        module = Module.concat([opts[:app] || Common.base_module(), "Component", component_module])
        Common.check_mod_name_validity!(module)
        Common.check_mod_name_availability!(module)
        generate(file_name, module)
```

Eiji7 commented on lib/mix/tasks/scenic.gen.scene.ex:
```diff
@@ -0,0 +1,73 @@
+      [] ->
+        Mix.Tasks.Help.run(["scenic.gen.scene"])
+
+      [scene_module_name | _] ->
+        Common.elixir_version_check!()
+        file_name = opts[:module] || Macro.underscore(scene_module_name)
+        mod_name = opts[:app] || scene_module_name
+        Common.check_mod_name_validity!(mod_name)
+        Common.check_mod_name_availability!(mod_name)
+
+        generate(file_name, mod_name)
```
```suggestion
      [scene_module | _] ->
        Common.elixir_version_check!()
        file_name = Macro.underscore(opts[:module] || scene_module)
        module = Module.concat([opts[:app] || Common.base_module(), "Scene", scene_module])
        Common.check_mod_name_validity!(module)
        Common.check_mod_name_availability!(module)
        generate(file_name, module)
```

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -0,0 +1,35 @@
+defmodule <%= @app_module_name %>.Components.<%= @component_module_name %> do  
```
```suggestion
defmodule <%= @component_module %> do  
```

Eiji7 commented on templates/gen_scene/lib/scenes/scene.ex.eex:
```diff
@@ -0,0 +1,40 @@
+defmodule <%= @app_module_name %>.Scene.<%= @scene_module_name %> do
```
```suggestion
defmodule <%= @scene_module %> do
```

Eiji7 commented on templates/gen_scene/lib/scenes/scene.ex.eex:
```diff
@@ -0,0 +1,40 @@
+  alias Scenic.ViewPort
+
+  import Scenic.Primitives
+  # import Scenic.Components
```
```suggestion
```

Eiji7 reviewed (changes requested):
Please try those changes.

crertel committed:
Update templates/gen_scene/lib/scenes/scene.ex.eex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update templates/gen_component/lib/components/component.ex.eex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update templates/gen_scene/lib/scenes/scene.ex.eex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.scene.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update templates/gen_component/lib/components/component.ex.eex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.scene.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update test/scenic_gen_scene_test.exs

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update test/scenic_gen_component_test.exs

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel commented:
(hold on, fixing some stuff up)

Eiji7 commented:
@crertel Thanks for adding those. Looks like you missed `3 hidden conversations`.

Eiji7 commented:
> @crertel (hold on, fixing some stuff up)
Not sure if you noticed, but you have missed `3 hidden conversations` above.

crertel committed:
Update templates/gen_component/lib/components/component.ex.eex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.scene.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel commented:
```
Created scene Elixir.ScenicNew.Scene.TestScene.
```
I'm seeing this show up elsewhere too...I think we're running into atom/module munging issues for the name. Posting up what I have to see if we can spot why/when this is happening.

crertel committed:
Fix atom-munging issues.

crertel commented:
Okay, that last commit should fix it. Test suite passes.

Eiji7 commented on lib/common.ex:
```diff
@@ -69,14 +69,22 @@ defmodule ScenicNew.Common do
     end
   end
 
+  def check_mod_name_validity!(name) when is_atom(name) do
+    unless Atom.to_string(name) =~ Regex.recompile!(~r/^[A-Z]\w*(\.[A-Z]\w*)*$/) do
+      Mix.raise(
+        "Module name must be a valid Elixir alias (for example: Foo.Bar), got: #{inspect(name)}"
+      )
+    end
+  end
```
This should not be needed, see:
https://github.com/phoenixframework/phoenix/blob/v1.5.7/lib/mix/tasks/phx.gen.presence.ex#L1

Eiji7 replied:
After accepting changes this is no longer used.
```suggestion
```

Eiji7 commented:
Heh, just noticed that … we don't need `opts[:module]` at all. :slightly_smiling_face: 

Going to add last suggestions soon …

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,60 @@
+  import Mix.Generator
+  alias ScenicNew.Common
+
+  @switches [app: :string, module: :string]
```
```suggestion
  @switches [app: :string]
```

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,60 @@
+
+      [component_module | _] ->
+        Common.elixir_version_check!()
+        file_name = Macro.underscore(opts[:module] || component_module)
+        module = Module.concat([opts[:app] || Common.base_module(), "Component", component_module])
+        Common.check_mod_name_validity!(module)
+        Common.check_mod_name_availability!(module)
+        generate(file_name, module)
```
```suggestion
        file_name = Macro.underscore(component_module)
        app_module = opts[:app] || Common.base_module()
        full_module = Module.concat([app_module, "Component", component_module])
        generate(file_name, component_module, full_module)
```

Eiji7 commented on lib/mix/tasks/scenic.gen.scene.ex:
```diff
@@ -0,0 +1,61 @@
+  import Mix.Generator
+  alias ScenicNew.Common
+
+  @switches [app: :string, module: :string]
```
```suggestion
  @switches [app: :string]
```

Eiji7 commented on lib/mix/tasks/scenic.gen.scene.ex:
```diff
@@ -0,0 +1,61 @@
+
+      [scene_module | _] ->
+        Common.elixir_version_check!()
+        file_name = Macro.underscore(opts[:module] || scene_module)
+        module = Module.concat([opts[:app] || Common.base_module(), "Scene", scene_module])
+
+        Common.check_mod_name_validity!(module)
+        Common.check_mod_name_availability!(module)
+        generate(file_name, module)
```
```suggestion
        file_name = Macro.underscore(scene_module)
        app_module = opts[:app] || Common.base_module()
        full_module = Module.concat([app_module, "Scene", scene_module])
        generate(file_name, scene_module, full_module)
```

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
```diff
@@ -0,0 +1,60 @@
+  end
+
+  # --------------------------------------------------------
+  defp generate(component_filename, component_module) do
+    component_path = "lib/components/#{component_filename}.ex"
+    create_directory("lib/components")
+    create_file(component_path, component_new_template(component_module: inspect(component_module)))
+    component_name = component_module |> Module.split() |> Enum.reverse() |> hd()
+    Mix.shell().info("Created component #{component_name}.")
```
```suggestion
  defp generate(component_filename, component_module, full_module) do
    component_path = "lib/components/#{component_filename}.ex"
    create_directory("lib/components")
    create_file(component_path, component_new_template(component_module: full_module))
    Mix.shell().info("Created component #{component_module}.")
```

Eiji7 commented on lib/mix/tasks/scenic.gen.scene.ex:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  # --------------------------------------------------------
+  defp generate(scene_filename, scene_module) do
+    scene_path = "lib/scenes/#{scene_filename}.ex"
+    create_directory("lib/scenes")
+    create_file(scene_path, scene_new_template(scene_module: inspect(scene_module)))
+    scene_name = scene_module |> Module.split() |> Enum.reverse() |> hd()
+    Mix.shell().info("Created scene #{scene_name}.")
```
```suggestion
  defp generate(scene_filename, scene_module, full_module) do
    scene_path = "lib/scenes/#{scene_filename}.ex"
    create_directory("lib/scenes")
    create_file(scene_path, scene_new_template(scene_module: full_module))
    Mix.shell().info("Created scene #{scene_module}.")
```

Eiji7 commented on templates/gen_component/lib/components/component.ex.eex:
```diff
@@ -0,0 +1,37 @@
+defmodule <%= @component_module %> do  
```
```suggestion
defmodule <%= inspect(@component_module) %> do  
```

Eiji7 commented on templates/gen_scene/lib/scenes/scene.ex.eex:
```diff
@@ -0,0 +1,39 @@
+defmodule <%= @scene_module %> do
```
```suggestion
defmodule <%= inspect(@scene_module) %> do
```

Eiji7 commented on test/scenic_gen_component_test.exs:
```diff
@@ -0,0 +1,64 @@
+  import ScenicNew.MixHelper
+  import ExUnit.CaptureIO
+
+  @component_name "test_component"
+  @component_module_name "TestComponent"
+  @app_name "scenic_new"
+  @module_name "ScenicNew"
+
+  test "gen.scene" do
+    func = fn ->
+      Mix.Tasks.Scenic.New.run([@app_name])
+      Mix.Tasks.Scenic.Gen.Component.run([@component_module_name])
+
+      assert_file("lib/components/#{@component_name}.ex", fn file ->
+        assert file =~ "defmodule #{@module_name}.Component.#{@component_module_name} do"
+      end)
+    end
+
+    in_tmp("new with defaults", fn ->
+      assert capture_io(func) =~ "Created component #{@component_module_name}."
```
```suggestion
  @app_module "ScenicNew"
  @app_name "scenic_new"
  @component_module "TestComponent"
  @component_name "test_component"

  test "gen.scene" do
    func = fn ->
      Mix.Tasks.Scenic.New.run([@app_name])
      Mix.Tasks.Scenic.Gen.Component.run([@component_module])

      assert_file("lib/components/#{@component_name}.ex", fn file ->
        assert file =~ "defmodule #{@app_module}.Component.#{@component_module} do"
      end)
    end

    in_tmp("new with defaults", fn ->
      assert capture_io(func) =~ "Created component #{@component_module}."
```

Eiji7 commented on test/scenic_gen_scene_test.exs:
```diff
@@ -0,0 +1,64 @@
+  import ScenicNew.MixHelper
+  import ExUnit.CaptureIO
+
+  @scene_name "test_scene"
+  @scene_module_name "TestScene"
+  @app_name "scenic_new"
+  @module_name "ScenicNew"
+
+  test "gen.scene" do
+    func = fn ->
+      Mix.Tasks.Scenic.New.run([@app_name])
+      Mix.Tasks.Scenic.Gen.Scene.run([@scene_module_name])
+
+      assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
+        assert file =~ "defmodule #{@module_name}.Scene.#{@scene_module_name} do"
+      end)
+    end
+
+    in_tmp("new with defaults", fn ->
+      assert capture_io(func) =~ "Created scene #{@scene_module_name}."
```
```suggestion
  @app_module "ScenicNew"
  @app_name "scenic_new"
  @scene_module "TestScene"
  @scene_name "test_scene"

  test "gen.scene" do
    func = fn ->
      Mix.Tasks.Scenic.New.run([@app_name])
      Mix.Tasks.Scenic.Gen.Scene.run([@scene_module])

      assert_file("lib/scenes/#{@scene_name}.ex", fn file ->
        assert file =~ "defmodule #{@app_module}.Scene.#{@scene_module} do"
      end)
    end

    in_tmp("new with defaults", fn ->
      assert capture_io(func) =~ "Created scene #{@scene_module}."
```

Eiji7 commented:
As mentioned here:
https://github.com/phoenixframework/phoenix/blob/v1.5.7/lib/mix/tasks/phx.gen.presence.ex#L1
There are no checks for `check_mod_name_validity` and `check_mod_name_availability`.

Secondly I passed value instead generating it like you do.

Remember that when doing: `"#{:atom}"` you need to inspect it like: `#{inspect(:atom)}` as otherwise for example for modules you would have `Elixir.` prefix.

Finally rest changes to pass formatter and tests.

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.scene.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.scene.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update templates/gen_scene/lib/scenes/scene.ex.eex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update test/scenic_gen_component_test.exs

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update test/scenic_gen_scene_test.exs

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel commented:
>  There are no checks for check_mod_name_validity and check_mod_name_availability.

Right, but note that that's how things are done in the other generators. I was trying to match the existing idioms of the codebase. I've merged your suggestions here, but my intent was to match Boyd's style.

crertel committed:
Update scenic.gen.component's generation to match scene's.

Eiji7 commented:
> @crertel Right, but note that that's how things are done in the other generators. I was trying to match the existing idioms of the codebase. I've merged your suggestions here, but my intent was to match Boyd's style.

Hmm … Those looks correct, but … for generating project! Look what `phx.new` is doing for it:
```elixir
    check_module_name_validity!(project.root_mod)
    check_module_name_availability!(project.root_mod)
 ```
If I understand intention correctly this is just to check if `app` you want to generate does not conflicts with already defined apps.

Let's test this code for our use case:
```elixir
defmodule Example do
  def check_mod_name_availability!(name) do
    [name]
    |> Module.concat()
    |> Module.split()
    |> Enum.reduce([], fn name, acc ->
      mod = Module.concat([Elixir, name | acc])
      # this interest us i.e. what would we check for
      IO.inspect(mod)
      # ensure loaded check called here
      # simplifying it to always pass
      [name | acc]
    end)
  end
end

iex> Example.check_mod_name_availability!("A.B.C")
A
B.A
C.B.A
["C", "B", "A"]
```
Do you think we really need those checks?

btw. Looks like you missed 3 suggestions again. :slightly_smiling_face: Could you please check them?

Eiji7 commented:
@crertel Looks like you fixed one of those 3 suggestions by hand. The only inconsistency between `component` and `scene` generation is a simple suggestion making `@switches` attribute to one line and a place where `inspect/1` is called (in one there is `generate/3` and in second it'a a template).

Eiji7 commented on lib/mix/tasks/scenic.gen.component.ex:
````diff
@@ -0,0 +1,58 @@
+  Generates Scene boilerplate for a new component.
+
+  ```bash
+  mix scenic.gen.component component_name
````
Oh, I completely forgot about documentation!
```suggestion
  mix scenic.gen.component ComponentName
```

Eiji7 commented on lib/mix/tasks/scenic.gen.scene.ex:
````diff
@@ -0,0 +1,58 @@
+  Generates Scene boilerplate for a new scene.
+
+  ```bash
+  mix scenic.gen.scene scene_name
````
Update documentation
```suggestion
  mix scenic.gen.scene SceneName
```

Eiji7 commented:
After those we can work on missing tests …

crertel committed:
Update lib/common.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.scene.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.component.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update lib/mix/tasks/scenic.gen.scene.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Update templates/gen_component/lib/components/component.ex.eex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel committed:
Fix excessive inspecting.

crertel committed:
Mix format.

Eiji7 commented:
It would be nice if you could add 2 extra changes:
```elixir
  def tmp_path do
    Path.expand("../tmp", __DIR__)
  end

  # …

  def in_tmp(which, function) do
    path = Path.join([tmp_path(), random_string(10), to_string(which)])

    try do
      File.rm_rf!(path)
      File.mkdir_p!(path)
      File.cd!(path, function)
    after
      path |> Path.dirname() |> File.rm_rf!()
    end
  end
```

The first in `tmp_path/0` returns path to `tmp` dir in `root` project directory instead of parent directory which does not makes sense.

Secondly `path |> Path.dirname() |> File.rm_rf!()` removes not only `Path.join([tmp_path(), random_string(10), to_string(which)])`, but also `Path.join([tmp_path(), random_string(10)])`. That way we would not have many empty directories in `tmp` directory.

Eiji7 commented:
Regarding tests here goes for `scene`:
```elixir
  @scene_module "TestScene"
  @scene_name "test_scene"

  test "gen.scene" do
    in_project("new with defaults", "my_app", fn _mix_project ->
      output = capture_io(fn -> Mix.Tasks.Scenic.Gen.Scene.run([@scene_module]) end)
      module_definition = "defmodule MyApp.Scene.#{@scene_module} do"
      assert_file("lib/scenes/#{@scene_name}.ex", module_definition)
      assert output =~ "Created scene #{@scene_module}."
    end)
  end

  test "gen.scene inside umbrella" do
    in_project("umbrella", "my_umbrella", ["--umbrella"], Mix.Tasks.New, fn _mix_project ->
      func = fn -> Mix.Tasks.Scenic.Gen.Scene.run([]) end
      assert_raise Mix.Error, func
    end)
  end

  test "gen.scene with app" do
    in_project("new with app", "my_app", ["--module", "My.App"], fn _mix_project ->
      args = [@scene_module, "--app", "My.App"]
      output = capture_io(fn -> Mix.Tasks.Scenic.Gen.Scene.run(args) end)
      assert output =~ "Created scene #{@scene_module}."
      module_definition = "defmodule My.App.Scene.#{@scene_module} do"
      assert_file("lib/scenes/#{@scene_name}.ex", module_definition)
    end)
  end

  test "gen.scene without args displays help" do
    in_project("help", "help", fn _mix_project ->
      output1 = capture_io(fn -> Mix.Tasks.Scenic.Gen.Scene.run([]) end)
      output2 = capture_io(fn -> Mix.Tasks.Help.run(["scenic.gen.scene"]) end)
      assert output1 == output2
    end)
  end

  defp in_project(tmp_dir, app_name, extra_args \\ [], task \\ Mix.Tasks.Scenic.New, project_func) do
    in_tmp(tmp_dir, fn ->
      capture_io(fn -> task.run([app_name | extra_args]) end)
      app_name |> String.to_atom() |> Mix.Project.in_project(app_name, project_func)
    end)
  end
```

Eiji7 commented:
And for `component`:
```elixir
  @component_module "TestComponent"
  @component_name "test_component"

  test "gen.component" do
    in_project("new with defaults", "my_app", fn _mix_project ->
      output = capture_io(fn -> Mix.Tasks.Scenic.Gen.Component.run([@component_module]) end)
      module_definition = "defmodule MyApp.Component.#{@component_module} do"
      assert_file("lib/components/#{@component_name}.ex", module_definition)
      assert output =~ "Created component #{@component_module}."
    end)
  end

  test "gen.component inside umbrella" do
    in_project("umbrella", "my_umbrella", ["--umbrella"], Mix.Tasks.New, fn _mix_project ->
      func = fn -> Mix.Tasks.Scenic.Gen.Component.run([]) end
      assert_raise Mix.Error, func
    end)
  end

  test "gen.component with app" do
    in_project("new with app", "my_app", ["--module", "My.App"], fn _mix_project ->
      args = [@component_module, "--app", "My.App"]
      output = capture_io(fn -> Mix.Tasks.Scenic.Gen.Component.run(args) end)
      assert output =~ "Created component #{@component_module}."
      module_definition = "defmodule My.App.Component.#{@component_module} do"
      assert_file("lib/components/#{@component_name}.ex", module_definition)
    end)
  end

  test "gen.component without args displays help" do
    in_project("help", "help", fn _mix_project ->
      output1 = capture_io(fn -> Mix.Tasks.Scenic.Gen.Component.run([]) end)
      output2 = capture_io(fn -> Mix.Tasks.Help.run(["scenic.gen.component"]) end)
      assert output1 == output2
    end)
  end

  defp in_project(tmp_dir, app_name, extra_args \\ [], task \\ Mix.Tasks.Scenic.New, project_func) do
    in_tmp(tmp_dir, fn ->
      capture_io(fn -> task.run([app_name | extra_args]) end)
      app_name |> String.to_atom() |> Mix.Project.in_project(app_name, project_func)
    end)
  end
  ```
  Look that `in_project/5` is same for both, so should go to `mix_helper.exs`.

crertel committed:
Cleanup testing, fix tmp folder creation.

crertel commented:
Okay, that should get us into a happier spot.

Eiji7 commented on lib/common.ex:
```diff
@@ -99,4 +99,14 @@ defmodule ScenicNew.Common do
     end
   end
+
+  def base_module, do: app_base(otp_app())
+  defp otp_app, do: Mix.Project.config() |> Keyword.fetch!(:app)
+
+  defp app_base(app) do
+    case Application.get_env(app, :namespace, app) do
+      ^app -> app |> to_string() |> Macro.camelize()
+      mod -> mod |> inspect()
+    end
+  end
```
I found this documentation:
https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Routes.html
Such configuration does not interest us, so we should remove it.

```suggestion
  def base_module, do: Mix.Project.config()[:app]
```

Eiji7 replied:
I found this documentation:
https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Routes.html
Such configuration does not interest us, so we should remove it.

Also this fixes last warning from `Codecov`:
```suggestion
  def base_module do
    Mix.Project.config() |> Keyword.fetch!(:app) |> Atom.to_string() |> Macro.camelize()
  end
```

Eiji7 commented:
After that remove `WIP` part from title as it would be ready to merge.

crertel committed:
Update lib/common.ex

Co-authored-by: Tomasz Marek Sulima <Eiji7@users.noreply.github.com>

crertel commented:
Okiedokie, let's get @boydm to take a look.

axelson committed:
Update to Scenic v0.11

Only update `new`, `gen_scene`, and `gen_component`

Switch to scenic_driver_local
Add a usage statement for gen_component
A few stylistic changes
Switch build_embedded to false to workaround a bug
Switch the scenic dep to the v0.11 branch (will need to be updated once
  a release is cut)

crertel committed:
Merge pull request #1 from axelson/update-to-scenic-0.11

Update to Scenic v0.11

Eiji7 commented:
@crertel This code is unfortunately invalid now as `v0.11` branch was deleted after merging PR. I guess we should mark it WIP until new `scenic` release, then change `scenic` version to latest from `hex` and finally mark it as ready to merge.

crertel committed:
Switch to use hex v0.11 candidate.

crertel committed:
Update test to check for correct scenic release.
